cobc/replace.c

```diff
@@ -0,0 +1,852 @@
+							    leading,
+							    src_text,text,
+							    new_text) ;
+
+				/* restart replacements on this stream */
+				check_replace_after_match (MORE_DEPTH repls);
+			} else {
+				check_replace (MORE_DEPTH repls,replace_list);
+			}
+		} else {
+			/* we need to compare a list of texts from
+			 * this stream with a list of texts from the
+			 * replacement */
+			check_replace_all (MORE_DEPTH repls,new_text,
+					   repls->token_queue,
+					   src->text_list,
+					   replace_list);
+		}
+	}
+}
+
+static COB_INLINE COB_A_INLINE int
+is_space_or_nl (const char c)
+{
+	return c == ' ' || c == '\n';
+}
+
+/* `check_replace_all( repls, new_text, texts, src, replace_list )`:
+ * checks whether a particular replacement is possible on the current
+ * list of texts.
+ * * `repls` is the current stream state
+ * * `new_text` is the text by which the texts should be replace in case of match
+ * * `texts` is the list of texts found in the source that remains to be matched
+ * * `src` is the list of texts from the replacement to be matched
+ * * `replace_list` is the next replacements to try in case of failure
+ */
+static
+void check_replace_all (WITH_DEPTH
+			struct cb_replacement_state *repls,
+			const struct cb_text_list *new_text,
+			struct cb_token_list *texts,
+			const struct cb_text_list *src,
+			const struct cb_replace_list *replace_list)
+{
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%scheck_replace_all(%s,",
+		DEPTH, repls->name);
+	fprintf(stderr, "%s    new_text = %s,\n", DEPTH,
+		string_of_text_list(new_text));
+	fprintf(stderr, "%s    texts = %s,\n", DEPTH,
+		string_of_token_list(texts));
+	fprintf(stderr, "%s    src = %s,\n", DEPTH,
+		string_of_text_list(src));
+	fprintf(stderr, "%s)\n", DEPTH);
+#endif
+
+	if (src==NULL){
+		/* MATCH */
+		/* pass the new text to the next stream */
+		ppecho_switch_text_list (MORE_DEPTH repls, new_text) ;
+		/* keep only in this stream the remaining texts that have not been matched */
+		repls->token_queue = texts ;
+		/* restart replacements on the stream */
+		check_replace_after_match (MORE_DEPTH repls);
+	} else {
+		const char* src_text = src->text;
+		if ( is_space_or_nl(src_text[0]) ){
+			/* skip spaces in replacement */
+			check_replace_all (MORE_DEPTH repls,new_text,texts,
+					   src->next, replace_list);
+		} else {
+			if (texts == NULL){
+				/* PARTIAL MATCH, we have emptied the
+				 * list of texts, but there are still
+				 * texts in the replacement, so wait
+				 * for more texts to be added on the
+				 * stream */
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%s  check_replace_all --> PARTIAL MATCH\n", DEPTH);
+#endif
+			} else {
+				const char* text = texts->text;
+				texts = texts->next;
+				if ( is_space_or_nl(text[0]) ){
+					/* skip spaces in texts */
+					check_replace_all (MORE_DEPTH repls,
+							   new_text,
+							   texts, src,
+							   replace_list);
+				} else {
+					if (!strcasecmp(src_text,text)){
```

Comment:
@lefessan We do have the text len(yylen) in the scanner, wouldn't it make sense to pass this and include it in the struct?
Then we can compare the length itself and don't need to use strcasecmp for all content (which is done as soon as at least one REPLACE[ING] is active, no?

Reply:
It's not obvious if the speedup would be worth the work. It's likely that, for most words, `strcasecmp(...)` immediately returns after checking only one or two chars. So, on one hand, testing the size would save some tests when the sizes are different, but on the other hand, if the sizes are equal, it adds one test that was not done before.

Reply:
Good point!